Panel/controllers/ad_controller.go

```diff
@@ -162,3 +163,59 @@ func (ctrl AdController) HandleEvent(c *gin.Context) {
 
 	c.JSON(http.StatusOK, gin.H{"message": "Event successfully processed"})
 }
+func (ctrl AdController) HandleEventAtomic(c *gin.Context) {
```

Comment:
Unimportant suggestion: You may want to use the adverb form: 'HandleEventAtomically', or change the adjective form to 'AtomicEventHandler'.